### ISSUE TITLE
Support compiling ARMv7l Architecture

### ISSUE THREAD

**Issue by xUnholy** [istio/istio#21094]
Given my architecture type is armv7l, and Kubernetes is compatible with armv7l architecture I would like to be able to compile Istio for this format. 

I've attempted to compile using the current build scripts however, they're not compiling for my 32bit arch, rather for 64bit which is incompatible. I've also run into issues where I had to get several of the build tools available to arm 32bit locally and that itself was a lot of work.

ARM 32bit should be supported, or at least possible to compile. I've currently got a Rasberry Pi 4b cluster running Kubernetes, I firmly believe ARM support should become more standardized in support.

**Comment by pisymbol** [comment]
I completely agree with @xUnholy as I too have a RPi 4B cluster waiting on Istio - I'd like to use k3sup to install it too!

**Comment by xUnholy** [comment]
@howardjohn 

Has there been any work to investigate building and supporting multiple architecture types using `docker buildx`? 

https://docs.docker.com/buildx/working-with-buildx/

In light of recently attempting to compile armv7l architecture arch myself, can I suggest the compilation of the source code be done with dependencies encapsulated in containers? This would make reproducible builds easier to troubleshoot and make version pinning extremely easy if there was a docker-compose like setup that the Makefiles execute. 

I've also engaged previously in another issue thread this discussion which I think might be handy linking if people like myself are seeking this kind of support.

Issue Ref: #13321

**Comment by howardjohn** [comment]
Building the docker images is the easier part, the harder part is getting Envoy compiled, which I don't think is (officially) done yet https://github.com/envoyproxy/envoy/issues/1861.

We are already using buildx, so getting it to do multi-arch could probably be done fairly easily

However, just because we can doesn't mean we should. We aren't going to ship something that isn't properly tested, documented, etc, which represents a huge amount of work - both short and long term.

**Comment by MrXinWang** [comment]
@xUnholy Referening my comments in your posted issue [here](https://github.com/raspbernetes/multi-arch-images/issues/5#issuecomment-598607763), personally I am quite interested in the multiarch support of istio, as istio has been already used in a lot of k8s-based projects.

@howardjohn Totally agree about your CI and document (seems much easier compared to the CI) concerns, but in order to make this multiarch official support (or arm support) a real thing, are there any suggestions from you that we can start to work on? I can contribute if the multiarch support stuff is welcomed by the community, and I think I will try to build and containerize these components on my arm64 machines first.

**Comment by howardjohn** [comment]
My opinion (and its just mine - others in the project may disagree) is that we would be perfectly happy to see changes made to help people build on various platforms more easily and accept PRs that move in this direction (assuming they don't have large complexity costs). To ship an "official" ARM image would be a step that I don't personally feel the project has the resources to adequately handle properly. I think it would make a lot of sense to have some community project building images in the short term and then we can evaluate making things more official in the future.

That's just my opinion though, you are welcome to join the test and release WG meeting (see https://github.com/istio/community/blob/master/WORKING-GROUPS.md#working-group-meetings) or even just add it to the agenda and we can discuss this with a wider audience.

**Comment by dave93cab** [comment]
how about just istioctl on ARMv7 and ARM64?

**Comment by howardjohn** [comment]
Yeah that is probably much easier, assuming no complications come up. Releases are built at istio/release-builder repo if anyone is interested in sending a PR. I think this would basically just involve a couple lines to add to the existing architectures

**Comment by dave93cab** [comment]
@howardjohn [https://github.com/istio/istio/pull/22381](https://github.com/istio/istio/pull/22381)

**Comment by xUnholy** [comment]
@dave93cab & @howardjohn I'm also in the process of not only making istioctl arm7l compliant, however, but it's also possible to run Istio in such a way that architecture is almost entirely irrelevant.

I am working on the Envoy support in parallel to this change and will hopefully update here soon with a corresponding pull request on their end.

https://github.com/istio/istio/pull/22455

**Comment by sdake** [comment]
@howardjohn regarding https://github.com/istio/istio/issues/21094#issuecomment-598787617. My opinion matches yours.

We should enable Istio's codebase to be built by third parties and distributed/tested by third parties for platforms we, ourselves, cannot CI (as a result of insufficient resources).

If we don't have a committed roadmap for testing the work, we shouldn't ship the containers directly in docker.io/istio.


Cheers
-steve

**Comment by dave93cab** [comment]
added https://github.com/istio/istio/pull/22567 for cleanup

**Comment by mmohamed** [comment]
Hi, I just finish with building Istio on ARM arch (v7/v8) and **it's not possible today**.

I have started by install all requirement for building environment (Golang, java, CMake, GCC, CC...)  and the first problem it to build Bazel from source (I try to use v0.28 with jdk8). Resolved, next problem is to build Istio proxy with Bazel from source, I got a problem with non-supported flag for GCC (old version) for envoy and pull container rule from docker rule , it's not compiled with ARM , so after compile it from source (with Bazel and go), I override it to make analyzing package success with Bazel, but I got new problem with ld linked and Ninja release (libac++ and libc++/abi not good liked). After resolving it, compile is started for great than 4000 packages (C, C++ ,Java and Go) but compilation was failed because envoy has a dependency to V8 chromium, and they have dependency to Wee8.

**Wee8 it's not available for ARM arch, so it's not possible to build Istio proxy with armv7 without re building Wee8  from source (if it’s possible) and force dependencies**.

_I don’t understand why it's so complicated to build a simple manageable proxy (envoy) and build a simple docker image for ARM. Mesh service it's a very simple concept but Istio with his more dependencies and complicated building toolchain (if I have understand it’s extends envoy to build core image) will be unmaintainable and can't be reused for another project._

_It's my opinion and I respect all works for making Istio. Before trying to build it from source, I have expecting Istio to be based on Nginx proxy who be injected inside pod to make a reverse manageable proxy and a simple frontend application with core back service to manage it._

**Thank you, I will try Istio if it's support ARM next time :)**

**Comment by morlay** [comment]
![image](https://user-images.githubusercontent.com/1667873/84335637-e100f780-abc7-11ea-93e1-3fc6a37d7e55.png)

Just share our build for arm64 https://github.com/querycap/istio （1.6.1 available too）

**hardest step is compiling istio envoy**. I pushed it here https://github.com/morlay/istio-envoy-arm64 (always timeout in github worflow https://github.com/morlay/istio-proxy-build-env/runs/759787862?check_suite_focus=true )

compiling istio enovy for arm64 is very slow.
16 hours cost on azure 4U16GiB x86_64 marchine in docker with qemu（`docker run --rm --privileged multiarch/qemu-user-static --reset -p yes`）

The build env here
https://github.com/morlay/istio-proxy-build-env/blob/master/build-env.Dockerfile

**two important changes**:

* added `gn` for  v8
* compiled `hsdis-<arch>.so` for bazel

And change env vars, then build envoy.
``` bash 
export PATH="${LLVM_PATH}/bin:${PATH}"; 
export JAVA_HOME="$(dirname $(dirname $(realpath $(which javac))))"; 
export BAZEL_BUILD_ARGS="--verbose_failures --define=ABSOLUTE_JAVABASE=${JAVA_HOME} --javabase=@bazel_tools//tools/jdk:absolute_javabase --host_javabase=@bazel_tools//tools/jdk:absolute_javabase --java_toolchain=@bazel_tools//tools/jdk:toolchain_vanilla --host_java_toolchain=@bazel_tools//tools/jdk:toolchain_vanilla"; 
make build_envoy; 
```
hope there be helpful.

**Comment by xUnholy** [comment]
@morlay This is fantastic news, I've also been interested in seeing this done for arm64 for some time. I raised a similar change into the upstream which you can view here for reference https://github.com/istio/istio/pull/22456

The outcome was they don't have the resources to be able to officially support building the arm64 / multi-arch images at this current stage, it would be awesome if we could potentially utalise the work you've done into the following repository as I know quite a number of people tracking this: https://github.com/raspbernetes/multi-arch-images

Issue Ref: https://github.com/raspbernetes/multi-arch-images/issues/5

**Comment by morlay** [comment]
@xUnholy welcome to did that.

but sadly. the compile times will block us to added to github workflow.

and for armv7, should fix the `gn` first. 
I just copy form https://github.com/envoyproxy/envoy-build-tools/blob/master/build_container/build_container_common.sh#L12-L27.

**Comment by xUnholy** [comment]
Github workflows should work with the accurately set timeout threshold being extended to the timeframe you've mentioned above. 

Would be great to see, also feel free to reach out, there is a discord chat link in the org label you can just and we can discuss further.

**Comment by hzxuzhonghu** [comment]
not stale

**Comment by tegk** [comment]
Envoy does now provice ARM64 builds https://www.envoyproxy.io/docs/envoy/latest/install/building#arm64-binaries so why not Istio?

**Comment by morlay** [comment]
@tegk 

i maintained a custom build of istio-envoy for arm64 https://github.com/querycap/istio-envoy-arm64 
You could download from the releases.

**Comment by ecliptik** [comment]
[Multi-arch support ](https://www.docker.com/blog/multi-arch-build-and-images-the-simple-way/)was merged into [distroless](https://github.com/GoogleContainerTools/distroless) today, https://github.com/GoogleContainerTools/distroless/pull/596, and there are now multi-arch `distroless` images in GCR which makes building Istio on arm64 much easier.

Using these steps, I was able to build Istio [1.7.3](https://github.com/istio/istio/releases/tag/1.7.3) and create [multi-arch images](https://www.docker.com/blog/multi-arch-build-and-images-the-simple-way/) for both arm64 and amd64.

Resulting multi-arch Images for amd64 and arm64:

[ecliptik/install-cni:1.7.3](https://hub.docker.com/r/ecliptik/install-cni/tags)
[ecliptik/operator:1.7.3](https://hub.docker.com/r/ecliptik/operator/tags)
[ecliptik/istioctl:1.7.3](https://hub.docker.com/r/ecliptik/istioctl/tags)
[ecliptik/mixer_codegen:1.7.3](https://hub.docker.com/r/ecliptik/mixer_codegen/tags)
[ecliptik/mixer:1.7.3](https://hub.docker.com/r/ecliptik/mixer/tags)
[ecliptik/test_policybackend:1.7.3](https://hub.docker.com/r/ecliptik/test_policybackend/tags)
[ecliptik/app_sidecar_centos_8:1.7.3](https://hub.docker.com/r/ecliptik/app_sidecar_centos_8/tags)
[ecliptik/app_sidecar_debian_10:1.7.3](https://hub.docker.com/r/ecliptik/app_sidecar_debian_10/tags)
[ecliptik/app_sidecar_debian_9:1.7.3](https://hub.docker.com/r/ecliptik/app_sidecar_debian_9/tags)
[ecliptik/app_sidecar_ubuntu_focal:1.7.3](https://hub.docker.com/r/ecliptik/app_sidecar_ubuntu_focal/tags)
[ecliptik/app_sidecar_ubuntu_bionic:1.7.3](https://hub.docker.com/r/ecliptik/app_sidecar_ubuntu_bionic/tags)
[ecliptik/app_sidecar_ubuntu_xenial:1.7.3](https://hub.docker.com/r/ecliptik/app_sidecar_ubuntu_xenial/tags)
[ecliptik/app:1.7.3](https://hub.docker.com/r/ecliptik/app/tags)
[ecliptik/proxyv2:1.7.3](https://hub.docker.com/r/ecliptik/proxyv2/tags)

## Building arm64 Istio Images 
These steps use the latest release of [1.7.3](https://github.com/istio/istio/releases/tag/1.7.3), but should also work with [master](https://github.com/istio/istio/tree/master/) and won't require the `git cherry-pick` in step 5.

1. Create a **a1.xlarge** EC2 instance on AWS with Ubuntu 20.04 and [setup the Istio dev environment](https://github.com/istio/istio/wiki/Preparing-for-Development) (docker, go, `~/.profile`, `$GOPATH`, `fpm`, `rpmbuild`, etc).

2. Clone and checkout tag `1.7.3`.

```
mkdir -p $HOME/go/src/istio.io/
cd $HOME/go/src/istio.io/
git clone https://github.com/istio/istio.git
cd istio
git checkout 1.7.3
```

3. Re-build the `istio/base` image locally so it's arm64 - otherwise it will pull from Dockerhub and use an x86 image and throw `Exec error format` errors.

```
docker build -f docker/Dockerfile.base -t istio/base:1.7-dev.7 .
```

4. Update `/pilot/docker/Dockerfile.proxyv2` to use the [distrocless:cc](https://github.com/GoogleContainerTools/distroless/tree/master/cc) multi-arch image SHA [ac75dbf0b249e5e3758134458d447bfd6a85d74c262ecb7cd88df68a3f53c6e3](https://console.cloud.google.com/gcr/images/distroless/GLOBAL/cc@sha256:ac75dbf0b249e5e3758134458d447bfd6a85d74c262ecb7cd88df68a3f53c6e3/details?tab=info) and `COPY` from `/usr/lib/aarch64-linux-gnu` instead of `/usr/lib/x86_64-linux-gnu`.

```
diff --git a/pilot/docker/Dockerfile.proxyv2 b/pilot/docker/Dockerfile.proxyv2
index a62ef00afc..8facece565 100644
--- a/pilot/docker/Dockerfile.proxyv2
+++ b/pilot/docker/Dockerfile.proxyv2
@@ -14,12 +14,12 @@ COPY gcp_envoy_bootstrap.json /var/lib/istio/envoy/gcp_envoy_bootstrap_tmpl.json
 RUN chown -R istio-proxy /var/lib/istio

 # The following section is used as base image if BASE_DISTRIBUTION=distroless
-FROM gcr.io/distroless/cc@sha256:f81e5db8287d66b012d874a6f7fea8da5b96d9cc509aa5a9b5d095a604d4bca1 as distroless
+FROM gcr.io/distroless/cc@sha256:ac75dbf0b249e5e3758134458d447bfd6a85d74c262ecb7cd88df68a3f53c6e3 as distroless

 # TODO(https://github.com/istio/istio/issues/17656) clean up this hack
 COPY --from=default /sbin/xtables-multi /sbin/iptables* /sbin/ip6tables* /sbin/ip /sbin/
-COPY --from=default /usr/lib/x86_64-linux-gnu/xtables/ /usr/lib/x86_64-linux-gnu/xtables
-COPY --from=default /usr/lib/x86_64-linux-gnu/ /usr/lib/x86_64-linux-gnu
+COPY --from=default /usr/lib/aarch64-linux-gnu/xtables/ /usr/lib/aarch64-linux-gnu/xtables
+COPY --from=default /usr/lib/aarch64-linux-gnu/ /usr/lib/aarch64-linux-gnu
 COPY --from=default /etc/iproute2 /etc/iproute2

 # Copy Envoy bootstrap templates used by pilot-agent
```

5. Cherry pick this commit, https://github.com/istio/istio/commit/f12b09c460694b5a107a061d4c323e4123bf87ba, to update the `Makefile.core.mk` and not hardcode `GOARCH` - will require resolving a small merge confict (skip this if building from `master`).

```
git cherry-pick -x f12b09c460694b5a107a061d4c323e4123bf87ba
```

6. Build Istio and containers (resulting image:tag are based off of env vars `HUB` and `TAG`).

```
BUILD_WITH_CONTAINER=0 make build
BUILD_WITH_CONTAINER=0 make docker
```

7. Inspect images and verify `Architecture` is `arm64`, eg for `ecliptik/proxyv2:1.7.3-arm64`.
```
docker image inspect ecliptik/proxyv2:1.7.3-arm64 | grep -i Architecture
        "Architecture": "arm64",
```

8. Run the image and verify it's kernel architecture is `aarch64`.

```
docker run -it --rm --entrypoint=uname ecliptik/proxyv2:1.7.3-arm64 -a      
Linux c5e1b77b5733 4.19.76-linuxkit #1 SMP Tue May 26 11:42:35 UTC 2020 aarch64 aarch64 aarch64 GNU/Linux
```

## Build a multi-arch image for amd64 and arm64

These steps borrow heavily from [Introducing multi-architecture container images for Amazon ECR](https://aws.amazon.com/blogs/containers/introducing-multi-architecture-container-images-for-amazon-ecr/), but should work with Docker Hub, ECR, and GCR as they all support [Image Manifest V 2, Schema 2](https://docs.docker.com/registry/spec/manifest-v2-2/).

1. Pull the corresponding amd64 image from Istio Dockerhub (these are currently not multi-arch and therefore only amd64)

```
docker pull istio/proxyv2:1.7.3
```

2. Retag image with personal repo and add `-amd64` suffix to the tag, then push it.

```
docker tag istio/proxyv2:1.7.3 ecliptik/proxyv2:1.7.3-amd64
docker push ecliptik/proxyv2:1.7.3-amd64
```

3. Create manifest using the -amd64 and -arm64 images.

```
docker manifest create ecliptik/proxyv2:1.7.3 ecliptik/proxyv2:1.7.3-amd64 ecliptik/proxyv2:1.7.3-arm64
docker manifest annotate --arch arm64 ecliptik/proxyv2:1.7.3 ecliptik/proxyv2:1.7.3-arm64
```

4. Push the manifest to the repo.

```
docker manifest push ecliptik/proxyv2:1.7.3
```

5. Verify image in the repo is multi-arch using [Docker Buildx](https://docs.docker.com/buildx/working-with-buildx/).

```
export DOCKER_CLI_EXPERIMENTAL=enabled

docker buildx imagetools inspect ecliptik/proxyv2:1.7.3
Name:      docker.io/ecliptik/proxyv2:1.7.3
MediaType: application/vnd.docker.distribution.manifest.list.v2+json
Digest:    sha256:d8f532846dd202a6e361c360e3abe404df0cbbd64fd4fc6d0f422ada51e1a88a

Manifests:
  Name:      docker.io/ecliptik/proxyv2:1.7.3@sha256:6169d096fe60f128f1311f76c97ee1c3e5d760a45042d5c9182492745d34c658
  MediaType: application/vnd.docker.distribution.manifest.v2+json
  Platform:  linux/amd64

  Name:      docker.io/ecliptik/proxyv2:1.7.3@sha256:b500dbf60174c516b158215e734b6d571a1d3c7d0ba3d625b814109a41552d25
  MediaType: application/vnd.docker.distribution.manifest.v2+json
  Platform:  linux/arm64
```

**Comment by morlay** [comment]
@ecliptik 

I notice you COPY same `envoy` binary for different arch in proxyv2.

```
COPY file:02e9d06202bbf4175f707c767b94191fdc2ff625f6f1c267703843238c2362b1 in /usr/local/bin/envoy 
```

arm64 version envoy binary need to compile from source https://github.com/istio/proxy (time consuming of step is very very long, 1 hour in pure 8vCPU 16GiB arm64 host)
it could download from https://github.com/querycap/istio-envoy-arm64/releases. 

I have simplify the build-env for envoy https://github.com/querycap/istio-envoy-arm64/blob/master/build-env/Dockerfile
next version we may use official build-env https://hub.docker.com/r/envoyproxy/envoy-build-ubuntu (now need to use llvm-9)

Glad to hear distroless land multi-arch support. I will add distroless images support later into https://github.com/querycap/istio.

**Comment by howardjohn** [comment]
Not stale

**Comment by ecliptik** [comment]
The Envoy 1.16 release notes include that ARM64 multi-arch images are officially supported and are available in Dockerhub

- https://www.envoyproxy.io/docs/envoy/v1.16.0/install/building#arm-binaries.

```
docker buildx imagetools inspect envoyproxy/envoy:v1.16.0

docker buildx imagetools inspect envoyproxy/envoy:v1.16.0                                   
Name:      docker.io/envoyproxy/envoy:v1.16.0
MediaType: application/vnd.docker.distribution.manifest.list.v2+json
Digest:    sha256:9e72bbba48041223ccf79ba81754b1bd84a67c6a1db8a9dbff77ea6fc1cb04ea
           
Manifests: 
  Name:      docker.io/envoyproxy/envoy:v1.16.0@sha256:0ab72c75ab1250475f83f5ce1adc11d8f7ecb9cd2a6bea419ffbf61bf154a0ed
  MediaType: application/vnd.docker.distribution.manifest.v2+json
  Platform:  linux/arm64
             
  Name:      docker.io/envoyproxy/envoy:v1.16.0@sha256:5e54002b16ad194cf1338a802fab9358ba7bd235360eb733b5871d1024219be4
  MediaType: application/vnd.docker.distribution.manifest.v2+json
  Platform:  linux/amd64
```

**Comment by ecliptik** [comment]
@morlay I see what you mean with the `envoy` binary for amd64 getting copied into the arm64 image. With envoy 1.16 supporting arm64 now, could you update the Dockerfile to use the `envoyproxy/envoy:v1.16.0` as a build stage, and copy it out into the istio image? Then you could re-use the upstream binary and not have to build it.

For example, something adding in something like this in `Dockerfile.proxyv2`,

```
FROM envoyproxy/envoy:v1.16.0 AS envoy 
COPY --from=envoy /usr/local/bin/envoy /usr/local/bin/envoy
```
Need to see if there are other binaries that would need a copy too, but it could be faster than having to re-compile `envoy`. It's also super hacky and not something that should go into the official Istio image either, but could be helpful as a workaround.

This is probably why PR https://github.com/istio/istio/pull/27743#issuecomment-709605136 is failing tests if `envoy` isn't an arm64 binary.

**Comment by howardjohn** [comment]
> Need to see if there are other binaries that would need a copy too, but
it could be faster than having to re-compile envoy. It's also super hacky
and not something that should go into the official Istio image either, but
could be helpful as a workaround.


Not sure if this was known or not, but the Istio build has custom
extensions compiled in. So you can play around with the upstream envoy one
but some things will not work for now. Long term we want to support
upstream envoy but we are a bit away from that.

**Comment by morlay** [comment]
@ecliptik 

istio use a fork from envoyproxy/enovywasm

which is not official envoyproxy/envoy

if you use official envoy, some features will broken

**Comment by BitRacer** [comment]
following.  I'm actually looking for arm64(v8) rpi cluster running microk8s.  Happy to help test when ready.

**Comment by dyaa** [comment]
Still, Not stale

**Comment by kittydoor** [comment]
This issue is not stale.

**Comment by howardjohn** [comment]
Agreed, sorry

**Comment by zackzhangkai** [comment]
Hi, how's going of istio arm binary now? Is there any plan on it?

**Comment by morlay** [comment]
https://github.com/istio/enhancements/issues/14

may be in plan, but i don't know details.

My custom build works well in my production environment.

**Comment by zackzhangkai** [comment]
@morlay  Thanks for your warning help. But I want to use the istio 1.6.10. I cannot found this arm image.

**Comment by morlay** [comment]
@zackzhangkai 
you could check old commit in the linked repo, and build by yourself.

i haven't enough time to maintain old versions.

**Comment by gabrielfreire** [comment]
Hi, any update on this ? 

Just bought a case and 3 raspberry PI 4 to transfer my cluster to it, installed everything, cluster up and running but istio won't install.

are the any workarounds at least ?

**Comment by chrstnwhlrt** [comment]
@gabrielfreire I think it should be possible to use istioctl to create a yaml containing the setup and just replace the images used with some arm64 compatible builds (from docker hub or self compiled)

**Comment by morlay** [comment]
@gabrielfreire could try to use my builds https://github.com/querycap/istio#install-istio

**Comment by gabrielfreire** [comment]
Thank you @morlay.

But i don't understand what i should do with this, where should i apply it ?
```yml
spec:
  components:
    pilot:
      k8s: # each components have to set this
        affinity: &affinity
          nodeAffinity:
            requiredDuringSchedulingIgnoredDuringExecution:
              nodeSelectorTerms:
                - matchExpressions:
                    - key: kubernetes.io/arch
                      operator: In
                      values:
                        - arm64
                        - amd64
    egressGateways:
      - k8s:
          affinity: *affinity
    ingressGateways:
      - k8s:
          affinity: *affinity
```

i believe i need it since i'm having this problem on ingress/egress
```
0/3 nodes are available: 3 node(s) didn't match Pod's node affinity.
```

**Comment by morlay** [comment]
@gabrielfreire 

using `1.10.x`，and just run

`istioctl operator init --hub=docker.io/querycapistio`

custom profiles should after you know more about istio-operator spec

the docs just show additional settings, not a full applyable crd yaml.

**Comment by gabrielfreire** [comment]
I am using `1.10.0`, just running doesn't work as i get the error: `0/3 nodes are available: 3 node(s) didn't match Pod's node affinity.`

I figured it out and added an **EDIT** to my previous post with the final istio-operator file that worked.

it seems that the key is not `kubernetes.io/arch` but `beta.kubernetes.io/arch`


Thanks for your work on this !

**Comment by morlay** [comment]
@gabrielfreire 

if you only have `beta.kubernetes.io/arch` label in node. 

Your k8s version may be too old？

**Comment by gabrielfreire** [comment]
@morlay 
```
 kubectl get no
NAME          STATUS   ROLES                      AGE   VERSION
k8s-node-01   Ready    controlplane,etcd,worker   23h   v1.20.6
k8s-node-02   Ready    controlplane,etcd,worker   23h   v1.20.6
k8s-node-03   Ready    controlplane,etcd,worker   23h   v1.20.6
```

**Comment by morlay** [comment]
@gabrielfreire 

Interesting. I remember k8s have multiple labels about arch since 1.13 or before.

Is your k8s created by cloud provider?

**Comment by gabrielfreire** [comment]
@morlay 
No, it's bare metal on raspberry-pi cluster with metallb as load balancer, rancher/rke, etc

i'm not using any cloud provider for anything at the moment

**Comment by aslanpour** [comment]
Is there a chance for us to use istio on ARM or not?!!

**Comment by morlay** [comment]
@aslanpour 

`aarch64` **yes** ! 
`aarch32` **no**. istio-proxy (envoy) depends chrome v8, which is not support 32bit.

**Comment by jocatalin** [comment]
I've switched to mac mini with m1 processor (arm based) as local development box. I can run istio in kind (via docker desktop) using helm charts. I didn't experienced any issues until now. There are some anti-affinity rules that need to be adjusted to make the chart work (add `arm64` as an architecture)

**Comment by mike-source** [comment]
@jocatalin Is there any chance you could elaborate on the steps you took/rules you adjusted exactly? I'm considering making the switch to a mac mini with m1, but need to be able to use istio.

Our dev setup is using minikube & docker at the moment...

**Comment by niteshkumarniranjan** [comment]
Recently we were facing the same issue, after some debugging and research I wrote this script,
It uses KinD but I think it can easily be replaceable with either `docker-desktop` or `minikube`

https://github.com/nowandme/k8s-istio-m1

**Comment by jxlwqq** [comment]
add `arm64` in the `nodeAffinity`:

```
kubectl patch deployments.apps \
  istio-ingressgateway \
  --namespace istio-system \
  --type='json' \
  -p='[
  {"op": "replace", "path": "/spec/template/spec/affinity/nodeAffinity/preferredDuringSchedulingIgnoredDuringExecution/0/preference/matchExpressions/0/values", "value": [amd64,arm64]},
  {"op": "replace", "path": "/spec/template/spec/affinity/nodeAffinity/requiredDuringSchedulingIgnoredDuringExecution/nodeSelectorTerms/0/matchExpressions/0/values", "value": [amd64,arm64,ppc64le,s390x]}
  ]'
 
kubectl patch deployments.apps \
  istio-egressgateway \
  --namespace istio-system \
  --type='json' \
  -p='[
  {"op": "replace", "path": "/spec/template/spec/affinity/nodeAffinity/preferredDuringSchedulingIgnoredDuringExecution/0/preference/matchExpressions/0/values", "value": [amd64,arm64]},
  {"op": "replace", "path": "/spec/template/spec/affinity/nodeAffinity/requiredDuringSchedulingIgnoredDuringExecution/nodeSelectorTerms/0/matchExpressions/0/values", "value": [amd64,arm64,ppc64le,s390x]}
  ]'
```

**Comment by sirAlexander** [comment]
> 
@jxlwqq Thanks for the update. With the above, Both Istio core and Istiod install ok but the Egress and Ingress gateways processing results in failure.

**Comment by jxlwqq** [comment]
> > 
> 
> @jxlwqq Thanks for the update. With the above, Both Istio core and Istiod install ok but the Egress and Ingress gateways processing results in failure.

open a new terminal, then run the patch command.

**Comment by KoukiMatsuda** [comment]
@jxlwqq 

Thank you for publishing the procedure.
I built the Istio environment by following the instructions you published. However, it did not work as expected.

I have built the Calico + Istio environment with the following configuration, and have performed the steps described in Istio.
However, the generation of the httpbin pod does not work with an error. 
Can you tell me what is going on?
https://istio.io/latest/docs/tasks/traffic-management/ingress/ingress-control/#configuring-ingress-using-an-istio-gateway


**[Environment]**

Desktop PC（Host）
　OS：Ubuntu 20.04.3 LTS 64bit

Raspberry Pi4（Client）
　OS：Ubuntu Server 20.04.3 LTS 64bit(arm64)


**[Prerequisite.]**

The following software must have been installed.
・Docker
・Kubernetes
・calicoctl

**[Implementation Procedure]**

Reference URL：https://docs.projectcalico.org/getting-started/kubernetes/quickstart

### ■Insall Calico

1．Create a Kubernetes cluster.

```
$ sudo kubeadm init --pod-network-cidr=192.168.0.0/16
$ mkdir -p $HOME/.kube
$ sudo cp -i /etc/kubernetes/admin.conf $HOME/.kube/config
$ sudo chown $(id -u):$(id -g) $HOME/.kube/config
```


2. Join a Woker node to the generated cluster.

`$ sudo kubeadm join <..>`


3．Deploy Calico.

```
$ kubectl create -f https://docs.projectcalico.org/manifests/tigera-operator.yaml
$ kubectl create -f https://docs.projectcalico.org/manifests/custom-resources.yaml
```


4．Verify that Calico pod has been activated.

`$ watch kubectl get pods -n calico-system`


5．Allow the Master node to create a Pod.

`$ kubectl taint nodes --all node-role.kubernetes.io/master-`


### ■Install Istio

　Reference URL：https://projectcalico.docs.tigera.io/security/app-layer-policy


1．Enable application layer policy

```
$calicoctl patch FelixConfiguration default --patch \
   '{"spec": {"policySyncPathPrefix": "/var/run/nodeagent"}}'
```

2．Install Istio Operator

`$istioctl operator init --hub=docker.io/querycapistio --tag=1.10.2`

3．Install Istio

```
$curl -L https://git.io/getLatestIstio | ISTIO_VERSION=1.10.2 sh -
$export PATH="$PATH:/home/ubuntu/istio-1.10.2/bin"
$istioctl install --set hub=docker.io/querycapistio --set profile=demo -y --set values.gateways.istio-ingressgateway.type=NodePort

$kubectl create -f - <<EOF
apiVersion: security.istio.io/v1beta1
kind: PeerAuthentication
metadata:
  name: default-strict-mode
  namespace: istio-system
spec:
  mtls:
    mode: STRICT
EOF

$ kubectl apply -f - <<EOF
apiVersion: install.istio.io/v1alpha1
kind: IstioOperator
metadata:
  namespace: istio-system
  name: example-istiocontrolplane
spec:
  hub: docker.io/querycapistio
  profile: demo
EOF
```

4．Add arm64 in the nodeAffinity

```
$kubectl patch deployments.apps \
  istio-ingressgateway \
  --namespace istio-system \
  --type='json' \
  -p='[
  {"op": "replace", "path": "/spec/template/spec/affinity/nodeAffinity/preferredDuringSchedulingIgnoredDuringExecution/0/preference/matchExpressions/0/values", "value": [amd64,arm64]},
  {"op": "replace", "path": "/spec/template/spec/affinity/nodeAffinity/requiredDuringSchedulingIgnoredDuringExecution/nodeSelectorTerms/0/matchExpressions/0/values", "value": [amd64,arm64,ppc64le,s390x]}
  ]'

$kubectl patch deployments.apps \
  istio-egressgateway \
  --namespace istio-system \
  --type='json' \
  -p='[
  {"op": "replace", "path": "/spec/template/spec/affinity/nodeAffinity/preferredDuringSchedulingIgnoredDuringExecution/0/preference/matchExpressions/0/values", "value": [amd64,arm64]},
  {"op": "replace", "path": "/spec/template/spec/affinity/nodeAffinity/requiredDuringSchedulingIgnoredDuringExecution/nodeSelectorTerms/0/matchExpressions/0/values", "value": [amd64,arm64,ppc64le,s390x]}
  ]'
```

5．Update Istio sidecar injector

```
$curl https://docs.projectcalico.org/manifests/alp/istio-inject-configmap-1.10.yaml -o istio-inject-configmap.yaml
$kubectl patch configmap -n istio-system istio-sidecar-injector --patch "$(cat istio-inject-configmap.yaml)"

```

6．Add Calico authorization services to the mesh

`$kubectl apply -f https://docs.projectcalico.org/manifests/alp/istio-app-layer-policy-envoy-v3.yaml`


7．Add namespace labels

`$kubectl label namespace default istio-injection=enabled`

**Comment by bastosvinicius** [comment]
Hello guys, is there still hope to make this feature available?

**Comment by ddelange** [comment]
Is this issue (`armv7l`) superseded by #26652 (`arm64`, found via https://github.com/resf/istio)?

**Comment by morlay** [comment]
@ddelange  

`armv7l` is 32bit processor, which chould not be supported because of `envoy`.

**Comment by howardjohn** [comment]
Lets close this in favor of https://github.com/istio/istio/issues/26652. `arm64` support is planned but `armv7l` is not
